### PR TITLE
#558321 design general purpose interfaces to empower reporting

### PR DESCRIPTION
**Pull request by eparovyshnaya**
- api bundle is provided

Signed-off-by: elena.parovyshnaya <elena.parovyshnaya@gmail.com>

### CODE REVIEW

File: bundles/org.eclipse.passage.loc.yars.api/OSGI-INF/l10n/bundle.properties [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#Properties file for org.eclipse.passage.lic.api
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/OSGI-INF/l10n/bundle.properties]: Please correct bundle symbolic name

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchedData.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.List;
+
+/**
+ * Aware of a particular {@linkplain Storage} structure, able to fetch data from
+ * it according to encapsulated business logic.
+ *
+ * @since 0.1
+ */
+@SuppressWarnings("unused")
+public interface FetchedData<B extends Storage<?>, R> {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchedData.java]: Do we really need "B" type parameter here?

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     ArSysOp - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.passage.loc.yars.internal.api;
+
+public interface ListMedia<T, C> {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java]: Please add javadoc
T and C type parameters should be documented

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	default ListMedia<T, C> start() {
+		return this;
+	}
+
+	default ListMedia<T, C> finish() {
+		return this;
+	}
+
+	default ListMedia<T, C> nodeStart(@SuppressWarnings("unused") T node) {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java]: maybe "startNode" ?

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchParams.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     ArSysOp - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.passage.loc.yars.internal.api;
+
+public interface FetchParams {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchParams.java]: Please add javadoc

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchedData.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ *******************************************************************************/
+package org.eclipse.passage.loc.yars.internal.api;
+
+import java.util.List;
+
+/**
+ * Aware of a particular {@linkplain Storage} structure, able to fetch data from
+ * it according to encapsulated business logic.
+ *
+ * @since 0.1
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/FetchedData.java]: Please add type parameter description

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Query.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     ArSysOp - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.passage.loc.yars.internal.api;
+
+public interface Query<S extends Storage<?>, R, P extends FetchParams> {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Query.java]: Please add javadoc

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Storage.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * the author desires.
+ * </p>
+ * <p>
+ * The only contract here is the type of core entity.
+ * </p>
+ *
+ * @since 0.1
+ */
+@SuppressWarnings("unused")
+public interface Storage<E> {
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Storage.java]: Something is wrong with unused parameters. Maybe we can add 
Class<E> storageType();
?

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java (hunk 3 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/ListMedia.java]: Lots of thing are to be documented here. Documentation iss going to follow in a fourth PR (after ref impls).

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Query.java (hunk 7 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Query.java]: Documentation is going to follow in a separate PR. Is it OK?

File: bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Storage.java (hunk 8 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.loc.yars.api/src/org/eclipse/passage/loc/yars/internal/api/Storage.java]: `E` is a part of contract here: we particularly say, that a storage is dedicated to a single type. 

The only reason for `E` to be unused is that there is no methods. It's just totally unclear what to demand of the storage, if we do not demand too much. 

Actual `E` value does not make sense at runtime, nut only to related interfaces signatures. So there will be no use in `storageType()` method.

**Comment by ruspl-afed** [comment]
@eparovyshnaya I resolved all javadoc-related conversations as we agreed to have it later

**Pull request merged**